sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala

```diff
@@ -3459,6 +3459,7 @@ case class GpuOverrides() extends Rule[SparkPlan] with Logging {
       } else {
         Seq.empty
       }
+      InputFileBlockRule.apply(wrap)
```

Comment:
We probably want to do the same thing with runAfterTagRules. And possibly just make this a part of those rules. They are doing essentially the same kind of thing.  Make sure that two different parts of the plan are both on the CPU or both on the GPU.

Reply:
Hi @revans2, This PR provides the rule to find the spark plan chain first, and then tag them not to run on GPU. runAfterTagRules seems can't satisfy the requirement.

Reply:
In order to do the transitions correctly we have rules about what you can, and cannot look at when tagging a part of the plan.

https://github.com/NVIDIA/spark-rapids/blob/f774e1a231416aab7cacc86f06010ef797fc19c7/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala#L293-L297

This makes it so one rule cannot change things in a way that will violate the goals of other rules.

`runAfterTagRules` is just a single location to run all of the rules that do not follow these requirements, so we can be 100% sure that they will not interfere with each other. 

https://github.com/NVIDIA/spark-rapids/blob/f774e1a231416aab7cacc86f06010ef797fc19c7/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala#L627-L643

For example right now `InputFileBlockRule` runs before `fixUpExchangeOverhead` in `runAfterTagRules`. If for some reason it ran after `fixUpExchangeOverhead` instead that would be a bug because it could force things onto the CPU that `fixUpExchangeOverhead` already looked at.

If we put them all together in one place we can reason about what the order should be and work to avoid these bugs early on.

Reply:
Thx for the explanation. Done